src/common/mempool.cc

```diff
@@ -12,21 +12,73 @@
  *
  */
 
+#include <thread>
 #include "include/mempool.h"
 #include "include/demangle.h"
 
-#if defined(_GNU_SOURCE) && defined(WITH_SEASTAR) && !defined(WITH_ALIEN)
-#else
-// Thread local variables should save index, not &shard[index],
-// because shard[] is defined in the class
-static thread_local size_t thread_shard_index = mempool::num_shards;
-#endif
-
 // default to debug_mode off
 bool mempool::debug_mode = false;
 
 // --------------------------------------------------------------
 
+namespace mempool {
+
+static size_t num_shard_bits;
+static size_t num_shards;
+
+static thread_local size_t thread_shard_index = max_shards;
+
+#if !defined(MEMPOOL_SCHED_GETCPU)
+// There is no cheap way of determining which core a thread is assigned to.
+// Statically assign threads to shards to minimize how many threads use the
+// same shard
+static int thread_shard_next;
+static std::mutex thread_shard_mtx;
+
+#else
+// Use sched_getcpu() to determine which core a thread is running on
+#endif
+
+}
+
+size_t mempool::get_num_shards(void) {
+  if (num_shards==0) {
+    size_t threads = std::thread::hardware_concurrency();
+    if (threads == 0) {
+      threads = default_shards;
+    }
+    if (threads < min_shards) {
+      threads = min_shards;
+    }
+    if (threads > max_shards) {
+      threads = max_shards;
+    }
+    threads--;
+    while (threads!=0) {
+      num_shard_bits++;
+      threads>>=1;
+    }
+    num_shards = 1 << num_shard_bits;
+  }
+  return num_shards;
+}
+
+int mempool::pick_a_shard_int(void) {
+#if !defined(MEMPOOL_SCHED_GETCPU)
+  if (thread_shard_index == max_shards) {
+    // Thread has not been assigned to a shard yet
+    std::lock_guard<std::mutex> lck (thread_shard_mtx);
+    thread_shard_index = thread_shard_next++ & ((1 << num_shard_bits) - 1);
+  }
+#else
+  if (thread_shard_index == max_shards) {
+    // Thread has not been assigned to a shard yet
+    thread_shard_index = sched_getcpu() & ((1 << num_shard_bits) - 1);
```

Comment:
This assumes seastar's reactors should _never_ change cores. I think they are pinned but I'm worried about the great quantifier. I don't know _when exactly_ this pinning happens and whether it's guaranteed this codes doesn't execute before.

In short: the complexity makes the thinking harder (and more painful for me ;-).

Reply:
Let me respond to my own comment: the periodical recheck should address this worry.

Still, the impression is _we're fighting the complexity with more complexity_.